tiledb/sm/array/array_directory.cc

```diff
@@ -408,6 +409,8 @@ bool ArrayDirectory::loaded() const {
 
 tuple<Status, optional<std::vector<URI>>> ArrayDirectory::list_root_dir_uris() {
   // List the array directory URIs
+  [[maybe_unused]] auto timer_se = stats_->start_timer("list_root_uris");
```

Comment:
Why the `[[maybe_unused]]` here, but not in array.cc?

Reply:
Ah yes! Thanks for opening this conversation, I should have probably raised the same question in Slack. The truth is I have no idea why I get the unused variable warning as error here and not anywhere else :confused: 
```
/Users/ypatia/TileDB/tiledb/sm/array/array_directory.cc:681:8: error: unused variable 'timer_se' [-Werror,-Wunused-variable]
  auto timer_se = stats_->start_timer("array_dir_compute_array_schema_uris");
  ```
Any ideas?

Reply:
Turns out these errors are thrown when compiling the array object library [1]. It turns out that the compilation of that object library doesn't correctly pass the `-DTILEDB_STATS` compiler definition, which results in the `stats_->start_timer("foo")` call returning an integer [2] which causes the unused variable warning. I pushed a possible fix that'll work for now [3], but it seems like we should probably be compiling all of those object libraries with the same settings we get from bootstrapping?

Out of curiosity I spent some time looking into how a compiler knows that RAII guards aren't unused and it turns out that the warning is suppressed if an object's constructor or destructor is non-default [3]. 

[1] https://github.com/TileDB-Inc/TileDB/blob/af4823aabda88f48243b0373ef126567d11f8436/tiledb/sm/array/CMakeLists.txt#L33-L36
[2] https://github.com/TileDB-Inc/TileDB/blob/de6632a91785277ad286c0b5b674f0e42fa331be/tiledb/sm/stats/stats.h#L89
[3] https://github.com/TileDB-Inc/TileDB/commit/2229d409ec44910a659d569834fa2df529f8d1a9
[4] https://gcc.gnu.org/bugzilla/show_bug.cgi?id=17006

Reply:
Story created: https://app.shortcut.com/tiledb-inc/story/24564

Reply:
thank you!